metrics/ci_integrations/lib/jenkins/adapter/jenkins_ci_client_adapter.dart

```diff
@@ -0,0 +1,182 @@
+  ///
+  /// The [limits] can be used to set a range-specifier for the request.
+  Future<JenkinsBuildingJob> _fetchBuilds(
+    String projectId, {
+    JenkinsQueryLimits limits = const JenkinsQueryLimits.empty(),
+  }) async {
+    final newBuildsFetchResult =
+        await jenkinsClient.fetchBuilds(projectId, limits: limits);
+    _throwIfInteractionUnsuccessful(newBuildsFetchResult);
+    return newBuildsFetchResult.result;
+  }
+
+  /// Checks the given [interactionResult] to be [InteractionResult.isSuccess].
+  ///
+  /// Throws [StateError] with the message of [interactionResult] if this
+  /// result is [InteractionResult.isError].
+  void _throwIfInteractionUnsuccessful(InteractionResult interactionResult) {
+    if (interactionResult.isError) {
+      throw StateError(interactionResult.message);
+    }
+  }
+
+  /// Processes the given [builds] to the list of [BuildData]s.
+  ///
+  /// The [jobName] is used to identify a building job for the builds.
+  /// The [startFromBuildNumber] is used to filter builds which number is less
+  /// than or equal to this value. This allows to avoid fetching old builds
+  /// since the range specifier in Jenkins API only provides ability to set
+  /// fetch limits but not to filter data to fetch.
+  Future<List<BuildData>> _processJenkinsBuilds(
+    List<JenkinsBuild> builds,
+    String jobName, {
+    int startFromBuildNumber,
+  }) {
+    final buildDataFutures = builds
+        .where((build) =>
+            _checkBuildFinishedAndInRange(build, startFromBuildNumber))
+        .map((build) => _mapJenkinsBuild(build, jobName));
+
+    return Future.wait(buildDataFutures);
+  }
+
+  /// Checks a [build] to be not [JenkinsBuild.building] and
+  /// satisfy a [minBuildNumber] parameter.
+  ///
+  /// The [minBuildNumber] is ignored if `null` or [num.isNegative].
+  bool _checkBuildFinishedAndInRange(JenkinsBuild build, int minBuildNumber) {
+    final buildNumberValid = minBuildNumber == null ||
+        minBuildNumber.isNegative ||
+        build.number > minBuildNumber;
+    return !build.building && buildNumberValid;
+  }
+
+  /// Maps [jenkinsBuild] to the [BuildData] instance with coverage
+  /// data fetching.
+  Future<BuildData> _mapJenkinsBuild(
+    JenkinsBuild jenkinsBuild,
+    String jobName,
+  ) async {
+    final coverageArtifact = jenkinsBuild.artifacts.firstWhere(
+      (artifact) => artifact.fileName == 'coverage-summary.json',
+      orElse: () => null,
+    );
+
+    CoverageJsonSummary coverage;
+
+    if (coverageArtifact != null) {
+      final artifactContentFetchResult =
+          await jenkinsClient.fetchArtifactByRelativePath(
+        jenkinsBuild.url,
+        coverageArtifact.relativePath,
+      );
+
+      _throwIfInteractionUnsuccessful(artifactContentFetchResult);
+
+      final artifactContent = artifactContentFetchResult.result;
+      coverage = CoverageJsonSummary.fromJson(artifactContent);
+    }
```

Comment:
I suggest we extract this code to a separate method and pass coverage to this _mapJenkinsBuild as a parameter due to SRP.